File: implementation/DotNet/Synapses/Model/NetElems/Neuron.fs
module Synapses.Model.Neuron

open FSharpx.Collections
open Synapses.Model
open Synapses.Model.NetElems
open System.Text.Json.Serialization

type Neuron = { activationF: Activation.Function
                weights: LazyList<float> }

[<JsonFSharpConverter>]
type SerializableNeuron =
    { activationF: Activation.SerializableFunction
      weights: List<float> }

let init (inputSize: int)
         (activationF: Activation.Function)
         (weightInitF: unit -> float)
         : Neuron =
    let weights = id
                  |> Seq.initInfinite
                  |> LazyList.ofSeq
                  |> LazyList.take (1 + inputSize)
                  |> LazyList.map
                        (fun _ -> weightInitF())
    { activationF = activationF
      weights = weights }

let output (input: LazyList<float>)
           (neuron: Neuron)
           : float =
    LazyList.cons 1.0 input
    |> Mathematics.dotProduct neuron.weights
    |> Activation.restrictedInput neuron.activationF
    |> neuron.activationF.f

let backPropagated
        (learnRate: float)
        (input: LazyList<float>)
        (outWithDiff: float * float)
        (neuron: Neuron)
        : LazyList<float> * Neuron =
    let (out, diff) = outWithDiff
    let common = out
                 |> neuron.activationF.inverse
                 |> neuron.activationF.deriv
                 |> (*) diff
    let inDiffs = LazyList.map ((*) common) input
    let newWeights = neuron.weights
                     |> LazyList.map2
                         (fun x w ->
                            w - learnRate * common * x
                         )
                         (LazyList.cons 1.0 input)
    let newNeuron = { neuron
                      with weights = newWeights }
    (inDiffs, newNeuron)

let serialized
        (neuron: Neuron):
        SerializableNeuron =
    { activationF =
          Activation.serialized neuron.activationF
      weights =
          LazyList.toList neuron.weights }

let deserialized
    (s: SerializableNeuron)
    : Neuron =
    { activationF =
          Activation.deserialized s.activationF
      weights =
          LazyList.ofList s.weights }

File: implementation/Scala/src/main/scala/synapses/model/netElems/Neuron.scala
package synapses.model.netElems

import scala.util.chaining._
import synapses.model.Mathematics
import synapses.model.netElems.Activation.ActivationSerialized

case class Neuron(activationF: Activation,
                  weights: LazyList[Double])

object Neuron {

  def init(inputSize: Int,
           activationF: Activation)
          (weightInitF: => Double)
  : Neuron = {
    val weights = LazyList
      .range(0, inputSize + 1)
      .map(_ => weightInitF)
    Neuron(activationF, weights)
  }

  def output(input: LazyList[Double])
            (neuron: Neuron)
  : Double = Mathematics
    .dotProduct(
      LazyList.cons(1.0, input),
      neuron.weights
    )
    .pipe(Activation.restrictedInput(neuron.activationF, _))
    .pipe(neuron.activationF.f)

  def backPropagated(learningRate: Double,
                     input: LazyList[Double],
                     outputWithError: (Double, Double))
                    (neuron: Neuron)
  : (LazyList[Double], Neuron) = {
    val (output, error) = outputWithError
    val outputInverse =
      neuron
        .activationF
        .inverse(output)
    val common =
      error * neuron.activationF.deriv(outputInverse)
    val inErrors = input.map(_ * common)
    val newWeights = neuron.weights
      .zip(LazyList.cons(1.0, input))
      .map { case (w, x) =>
        w - learningRate * common * x
      }
    val newNeuron = neuron.copy(weights = newWeights)
    (inErrors, newNeuron)
  }

  case class NeuronSerialized(activationF: ActivationSerialized,
                              weights: List[Double])

  def serialized(neuron: Neuron): NeuronSerialized =
    NeuronSerialized(
      Activation.serialized(neuron.activationF),
      neuron.weights.toList
    )

  def deserialized(neuronSerialized: NeuronSerialized): Neuron =
    Neuron(
      Activation.deserialized(neuronSerialized.activationF),
      neuronSerialized.weights.to(LazyList)
    )

}
